src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -3,45 +3,418 @@
+            /// Partially decode field signature to obtain CorElementType and optionally the type handle for VALUETYPE fields.
+            /// </summary>
+            /// <param name="fieldDef">Metadata field definition</param>
+            /// <param name="metadataReader">Metadata reader for the active module</param>
+            /// <param name="corElementType">Output element type decoded from the signature</param>
+            /// <param name="valueTypeHandle">Value type handle decoded from the signature</param>
+            private static void GetFieldElementTypeAndValueTypeHandle(
+                in FieldDefinition fieldDef,
+                MetadataReader metadataReader,
+                out CorElementType corElementType,
+                out Handle valueTypeHandle)
+            {
+                ImmutableArray<byte> signature = metadataReader.GetBlobContent(fieldDef.Signature);
+                int index = 0;
+                SignatureHeader signatureHeader = new SignatureHeader(signature[index++]);
+                if (signatureHeader.Kind != SignatureKind.Field)
+                {
+                    throw new InvalidProgramException();
+                }
+
+                corElementType = ReadElementType(signature, ref index);
+                valueTypeHandle = default(Handle);
+                if (corElementType == CorElementType.ELEMENT_TYPE_GENERICINST)
+                {
+                    corElementType = ReadElementType(signature, ref index);
+                }
+
+                if (corElementType == CorElementType.ELEMENT_TYPE_VALUETYPE)
+                {
+                    valueTypeHandle = ReadCompressedToken(signature, ref index);
+                }
+            }
+
+            /// <summary>
+            /// Extract element type from a field signature after skipping various modifiers.
+            /// </summary>
+            /// <param name="signature">Signature byte array</param>
+            /// <param name="index">On input, index into the signature array. Gets modified to point after the element type on return.</param>
+            /// <returns></returns>
+            private static CorElementType ReadElementType(ImmutableArray<byte> signature, ref int index)
+            {
+                // SigParser::PeekElemType
+                byte signatureByte = signature[index];
+                if (signatureByte < (byte)CorElementType.ELEMENT_TYPE_CMOD_REQD)
+                {
+                    // Fast path
+                    index++;
+                    return (CorElementType)signatureByte;
+                }
+
+                // SigParser::SkipCustomModifiers -> SkipAnyVASentinel
+                if (signatureByte == (byte)CorElementType.ELEMENT_TYPE_SENTINEL)
+                {
+                    signatureByte = signature[++index];
+                }
+
+                // SigParser::SkipCustomModifiers - modifier loop
+                while (signatureByte == (byte)CorElementType.ELEMENT_TYPE_CMOD_REQD ||
```

Comment:
To test the modifier handling, could you throw a volatile static field into the test? That should be emitted as a field with a modifier.

Reply:
Added.